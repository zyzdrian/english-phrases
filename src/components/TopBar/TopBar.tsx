import React, { useState } from 'react';
import { Link } from '@reach/router';
import styled from 'styled-components';
import hamburger from './menu.svg';
import { FadeAnimation, Portal } from '..';

export enum TOP_BAR_TEST_IDS {
    title = 'title',
    sideBar = 'side-bar',
    hamburger = 'hamburger',
    backdrop = 'backdrop'
}

const StyledProgressBar = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: sticky;
  width: 100%;
  top: 0;
  padding: 0 24px;
  margin-bottom: 24px;
  height: 56px;
  background-color: ${props => props.theme.primary};
  transition: background-color ease 0.3s;
`;

const StyledHamburger = styled.img`
  cursor: pointer;
`;

const StyledText = styled.div`
font-size: 1.2rem;
`;

interface IStyledSideBar {
    isOpen: boolean;
}

const SIDE_BAR_POSITION = {
    open: 0,
    close: -200
};

const StyledSideBar = styled.div<IStyledSideBar>`
position: absolute;
width: 200px;
left: ${props => props.isOpen ? SIDE_BAR_POSITION.open : SIDE_BAR_POSITION.close}px;
top: 56px;
height: calc(100vh - 56px);
background-color: ${props => props.theme.primary};
display: flex;
flex-direction: column;
transition: left 0.3s ease-in-out;
z-index: 1;

& a {
  background-color: ${props => props.theme.primary};
  text-decoration: none;
  color: ${props => props.theme.fontColorLight};
  padding: 16px;
  transition: background-color 0.3s ease;
  
  &:first-child {
  margin-top: 16px;
  }
  
  &:hover {
    background-color: ${props => props.theme.secondary};
  }
}
`;

const StyledBackdrop = styled.div`
  background-color: rgba(0, 0, 0, 0.3);
  height: calc(100vh - 56px);
  width: 100vw;
  position: absolute;
  top: 56px;
  left: 0;
`;

export const TopBar = ({ title }: IProgressBarProps) => {
    const [isOpen, setIsOpen] = useState<boolean>(false);
    return (
        <StyledProgressBar>
            <StyledHamburger
                src={hamburger}
                alt='menu'
                onClick={() => setIsOpen(prevState => !prevState)}
                data-testid={TOP_BAR_TEST_IDS.hamburger}
            />
            <StyledText data-testid={TOP_BAR_TEST_IDS.title}>{title}</StyledText>
            <span style={{ width: '24px' }}/>
            <Portal id='side-bar'>
                <StyledSideBar isOpen={isOpen} data-testid={TOP_BAR_TEST_IDS.sideBar}>
                    <Link to='/'>Flashcards</Link>
                    <Link to='/free-learning'>Free learning</Link>
                </StyledSideBar>
                <FadeAnimation isOpen={isOpen}>
                    <StyledBackdrop onClick={() => setIsOpen(false)} data-testid={TOP_BAR_TEST_IDS.backdrop} />
                </FadeAnimation>
            </Portal>
        </StyledProgressBar>
    );
};

export interface IProgressBarProps {
    title?: string;
}
