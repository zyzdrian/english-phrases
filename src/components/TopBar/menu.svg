<svg xmlns="http://www.w3.org/2000/svg" width="24" height="16.646" viewBox="0 0 24 16.646">
    <defs>
        <style>
            .cls-1{fill:#fff}
        </style>
    </defs>
    <g id="menu" transform="translate(0 -78.439)">
        <g id="Group_2" data-name="Group 2" transform="translate(0 85.78)">
            <g id="Group_1" data-name="Group 1">
                <path id="Path_1" d="M23.031 235.318H.969a.982.982 0 0 0 0 1.964h22.062a.982.982 0 0 0 0-1.964z" class="cls-1" data-name="Path 1" transform="translate(0 -235.318)"/>
            </g>
        </g>
        <g id="Group_4" data-name="Group 4" transform="translate(0 78.439)">
            <g id="Group_3" data-name="Group 3">
                <path id="Path_2" d="M23.031 78.439H.969a.982.982 0 0 0 0 1.964h22.062a.982.982 0 0 0 0-1.964z" class="cls-1" data-name="Path 2" transform="translate(0 -78.439)"/>
            </g>
        </g>
        <g id="Group_6" data-name="Group 6" transform="translate(0 93.122)">
            <g id="Group_5" data-name="Group 5">
                <path id="Path_3" d="M23.031 392.2H.969a.982.982 0 0 0 0 1.964h22.062a.982.982 0 0 0 0-1.964z" class="cls-1" data-name="Path 3" transform="translate(0 -392.197)"/>
            </g>
        </g>
    </g>
</svg>
