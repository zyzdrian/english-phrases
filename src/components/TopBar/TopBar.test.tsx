import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { TopBar, IProgressBarProps, TOP_BAR_TEST_IDS } from './TopBar';
import { renderTheme } from '../../styledComponents';

describe('<FlipCard /> tests', () => {
    const INITIAL_PROPS: IProgressBarProps = {};
    let props = { ...INITIAL_PROPS };
    beforeEach(() => {
        props = { ...INITIAL_PROPS };
    });
    afterEach(cleanup);

    it('should render as default', () => {
        const { container, getByTestId } = renderTheme(<TopBar {...props} />);
        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId(TOP_BAR_TEST_IDS.sideBar)).toHaveStyle('left: -200px');
    });

    it('should render title', () => {
        props.title = 'Test title';

        const { container, getByTestId } = render(<TopBar {...props} />);
        expect(container.firstChild).toMatchSnapshot();
        expect(getByTestId(TOP_BAR_TEST_IDS.title)).toHaveTextContent(props.title);
    });

    it('should show sideBar when hamburger icon was clicked', () => {
        const { getByTestId } = renderTheme(<TopBar {...props} />);
        fireEvent.click(getByTestId(TOP_BAR_TEST_IDS.hamburger));

        expect(getByTestId(TOP_BAR_TEST_IDS.sideBar)).toHaveStyle('left: 0px');
    });

    it('should hide sideBar when hamburger icon was clicked two times', () => {
        const { getByTestId } = renderTheme(<TopBar {...props} />);
        fireEvent.click(getByTestId(TOP_BAR_TEST_IDS.hamburger));
        fireEvent.click(getByTestId(TOP_BAR_TEST_IDS.hamburger));

        expect(getByTestId(TOP_BAR_TEST_IDS.sideBar)).toHaveStyle('left: -200px');
    });

    it('should hide sideBar when backdrop was clicked', () => {
        const { getByTestId } = renderTheme(<TopBar {...props} />);
        fireEvent.click(getByTestId(TOP_BAR_TEST_IDS.hamburger));
        fireEvent.click(getByTestId(TOP_BAR_TEST_IDS.backdrop));

        expect(getByTestId(TOP_BAR_TEST_IDS.sideBar)).toHaveStyle('left: -200px');
    });
});
