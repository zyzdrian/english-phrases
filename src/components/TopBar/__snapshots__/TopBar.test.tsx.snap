// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlipCard /> tests should render as default 1`] = `
<div
  class="sc-VigVT AXRul"
>
  <img
    alt="menu"
    class="sc-jTzLTM kMqZix"
    data-testid="hamburger"
    src="menu.svg"
  />
  <div
    class="sc-fjdhpX jmFvEE"
    data-testid="title"
  />
  <span
    style="width: 24px;"
  />
</div>
`;

exports[`<FlipCard /> tests should render title 1`] = `
<div
  class="sc-VigVT cybuwg"
>
  <img
    alt="menu"
    class="sc-jTzLTM kMqZix"
    data-testid="hamburger"
    src="menu.svg"
  />
  <div
    class="sc-fjdhpX jmFvEE"
    data-testid="title"
  >
    Test title
  </div>
  <span
    style="width: 24px;"
  />
</div>
`;
